cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringProgramEventPublisher.java

```diff
@@ -0,0 +1,232 @@
+      lastProgramUpdateMessageId = appMetadataStore.retrieveSubscriberState(programUpdateTopic,
+                                                                            TetheringAgentService.SUBSCRIBER);
+      if (lastProgramUpdateMessageId == null) {
+        // Initialize subscriber based on last message fetched by RuntimeProgramStatusSubscriberService.
+        // This is to avoid fetching existing program history from before TetheringAgent was added
+        String messageId = appMetadataStore.retrieveSubscriberState(programUpdateTopic,
+                                                                    RuntimeProgramStatusSubscriberService.SUBSCRIBER);
+        appMetadataStore.persistSubscriberState(programUpdateTopic, TetheringAgentService.SUBSCRIBER, messageId);
+        lastProgramUpdateMessageId = messageId;
+      }
+    });
+  }
+
+  @Override
+  protected long runTask() throws IOException {
+    List<PeerInfo> peers = store.getPeers().stream()
+      // Ignore peers in REJECTED state.
+      .filter(p -> p.getTetheringStatus() != TetheringStatus.REJECTED)
+      .collect(Collectors.toList());
+
+    PeerProgramUpdates peerProgramUpdates = getPeerProgramUpdates();
+    persistNotifications(peerProgramUpdates, peers);
+    TransactionRunners.run(transactionRunner, context -> {
+      AppMetadataStore appMetadataStore = AppMetadataStore.create(context);
+      appMetadataStore.persistSubscriberState(programUpdateTopic,
+                                              TetheringAgentService.SUBSCRIBER,
+                                              lastProgramUpdateMessageId);
+
+    });
+
+    return pollInterval;
+  }
+
+  /**
+   * Returns program status updates for tethered peers.
+   */
+  private PeerProgramUpdates getPeerProgramUpdates() {
+    Map<String, List<Notification>> peerToNotifications = new HashMap<>();
+    String lastMessageId = lastProgramUpdateMessageId;
+    try (CloseableIterator<Message> iterator = messageFetcher.fetch(NamespaceId.SYSTEM.getNamespace(),
+                                                                    programUpdateTopic,
+                                                                    programUpdateFetchSize,
+                                                                    lastProgramUpdateMessageId)) {
+      while (iterator.hasNext()) {
+        Message message = iterator.next();
+        Notification notification = message.decodePayload(r -> GSON.fromJson(r, Notification.class));
+        if (notification.getNotificationType() == Notification.Type.PROGRAM_STATUS) {
+          Map<String, String> properties = notification.getProperties();
+          String programRunId = properties.get(ProgramOptionConstants.PROGRAM_RUN_ID);
+          try {
+            RunRecord runRecord = runRecordFetcher.getRunRecordMeta(GSON.fromJson(programRunId, ProgramRunId.class));
+            if (runRecord.getPeerName() != null) {
```

Comment:
this method has too much nesting. It's better to invert checks like this:

```
while (iterator.hasNext()) {
  ...
  if (notification.getType() != PROGRAM_STATUS) {
    continue;
  }
  try {
    RunRecord runRecord = ...;
    if (runRecord.getPeerName() == null) {
      continue;
    }
    ...
  }
}
```

Reply:
fixed